Lib/subprocess.py

```diff
@@ -968,12 +972,40 @@ def _execute_child(self, args, executable, preexec_fn, close_fds,
             # Process startup details
             if startupinfo is None:
                 startupinfo = STARTUPINFO()
-            if -1 not in (p2cread, c2pwrite, errwrite):
+
+            use_std_handles = -1 not in (p2cread, c2pwrite, errwrite)
+            if use_std_handles:
                 startupinfo.dwFlags |= _winapi.STARTF_USESTDHANDLES
                 startupinfo.hStdInput = p2cread
                 startupinfo.hStdOutput = c2pwrite
                 startupinfo.hStdError = errwrite
 
+            attribute_list = startupinfo.lpAttributeList
+            have_handle_list = (attribute_list and
+                                "handle_list" in attribute_list and
+                                attribute_list["handle_list"])
+
+            # If we were given an handle_list or need to create one
+            if have_handle_list or (use_std_handles and close_fds):
+                if attribute_list is None:
+                    attribute_list = startupinfo.lpAttributeList = {}
+                handle_list = attribute_list.setdefault("handle_list", [])
+
+                if use_std_handles:
+                    handle_list += [int(p2cread), int(c2pwrite), int(errwrite)]
+
+                handle_list[:] = self._filter_handle_list(handle_list)
+
+                if handle_list:
+                    if have_handle_list and not close_fds:
```

Comment:
At the level of the subprocess API, `close_fds` means the child will not inherit all of the inheritable handles of the current process, except for the standard handles, and if `close_fds` is false, it means all inheritable handles will be inherited. Using an explicit `handle_list` inherits only a given subset of handles and, like `pass_fds`, it's acting like an extension of the standard set of 3 handles  or file descriptors. As such, it should be used with `close_fds` as true. For `pass_fds` a warning is raised when `close_fds` is false because it's a mixed message and the code is opting to override `close_fds`. For consistency, using `handle_list` with `close_fds` as false should raise a similar warning. I know it's different in the Windows API, which instead modifies what bInheritHandles means when given a handle list. But I think the goal in `subprocess.Popen` is try to make the semantics as universal as possible. `pass_fds` and `handle_list` are obviously different, but they share the semantics of the common `close_fds` parameter.

As far as I can see -- both from reading and testing -- removing `have_handle_list` from the `if` statement should never cause a warning to be raised when only redirecting the standard handles. If there's no explicitly passed `handle_list`, this code block executes only when `close_fds` is true. So if you're checking `not close_fds` to gate the warning, there's no way it will be raised. If `close_fds` is false when redirecting the standard handles, normal inheritance of the entire process handle table is used -- unless of course a `handle_list` is already present.